src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs

```diff
@@ -87,16 +144,154 @@ internal HttpMessageHandler CreateHandler(string name)
+            var expired = new ExpiredHandlerTrackingEntry(active);
+            _expiredHandlers.Enqueue(expired);
+
+            StartCleanupTimer();
+        }
+
+        // Internal so it can be overridden in tests
+        internal virtual void StartCleanupTimer()
+        {
+            lock (_cleanupTimerLock)
+            {
+                if (_cleanupTimer == null)
+                {
+                    _cleanupTimer = new Timer(_cleanupCallback, null, DefaultCleanupInterval, Timeout.InfiniteTimeSpan);
+                }
+            }
+        }
+
+        // Internal for tests
+        internal void CleanupTimer_Tick(object state)
+        {
+            // Stop any pending timers, we'll restart the timer if there's anything left to process after cleanup.
+            //
+            // With the scheme we're using it's possible we could end up with some redundant cleanup operations.
+            // This is expected and fine.
+            // 
+            // An alternative would be to take a lock during the whole cleanup process. This isn't ideal because it
+            // would result in threads executing ExpiryTimer_Tick as they would need to block on cleanup to figure out
+            // whether we need to start the timer.
+            lock (_cleanupTimerLock)
+            {
+                _cleanupTimer.Dispose();
+                _cleanupTimer = null;
+            }
+
+            var stopwatch = ValueStopwatch.StartNew();
+
+            var initialCount = _expiredHandlers.Count;
+            Log.CleanupCycleStart(_logger, initialCount);
+
+            var disposedCount = 0;
+            for (var i = 0; i < initialCount; i++)
+            {
+                // Since we're the only one removing from _expired, TryDequeue must always succeed.
+                _expiredHandlers.TryDequeue(out var entry);
+                Debug.Assert(entry != null);
+
+                if (entry.CanDispose)
+                {
+                    try
+                    {
+                        entry.InnerHandler.Dispose();
+                        disposedCount++;
+                    }
+                    catch (Exception ex)
+                    {
+                        Log.CleanupItemFailed(_logger, entry.Name, ex);
+                    }
+                }
+                else
+                {
+                    // If the entry is still live, put it back in the queue so we can process it 
+                    // during the next cleanup cycle.
+                    _expiredHandlers.Enqueue(entry);
+                }
+            }
+
+            var remainingCount = _expiredHandlers.Count;
+            Log.CleanupCycleEnd(_logger, stopwatch.GetElapsedTime(), disposedCount, remainingCount);
+
+            // We didn't totally empty the cleanup queue, try again later.
+            if (remainingCount > 0)
+            {
+                StartCleanupTimer();
+            }
+        }
+
+        private static class Log
```

Comment:
I assume you'll pull this out in the non-design version.

Reply:
No this is the pattern we're using. It's much cleaner than a single class with infinity methods

Reply:
@BrennanConroy we should do this in SignalR.

Reply:
I know, you made an issue for it, after preview1